plonky2/examples/square_root.rs

```diff
@@ -0,0 +1,84 @@
+use std::marker::PhantomData;
+
+use anyhow::Result;
+use plonky2::field::types::{Field, PrimeField};
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::generator::{GeneratedValues, SimpleGenerator};
+use plonky2::iop::target::Target;
+use plonky2::iop::witness::{PartialWitness, PartitionWitness, Witness};
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use plonky2::plonk::circuit_data::CircuitConfig;
+use plonky2::plonk::config::{GenericConfig, PoseidonGoldilocksConfig};
+use plonky2_field::extension::Extendable;
+
+/// A generator used by the prover to calculate the square root (`x`) of a given value
+/// (`x_squared`), outside of the circuit, in order to supply it as an additional public input.
+#[derive(Debug)]
+struct SquareRootGenerator<F: RichField + Extendable<D>, const D: usize> {
+    x: Target,
+    x_squared: Target,
+    _phantom: PhantomData<F>,
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> SimpleGenerator<F>
+    for SquareRootGenerator<F, D>
+{
+    fn dependencies(&self) -> Vec<Target> {
+        vec![self.x_squared]
+    }
+
+    fn run_once(&self, witness: &PartitionWitness<F>, out_buffer: &mut GeneratedValues<F>) {
+        let x_squared = witness.get_target(self.x_squared);
+        let x = x_squared.sqrt().unwrap();
+
+        out_buffer.set_target(self.x, x);
+    }
+}
+
+/// An example of using Plonky2 to prove a statement of the form
+/// "I know the square root of this field element."
+fn main() -> Result<()> {
+    const D: usize = 2;
+    type C = PoseidonGoldilocksConfig;
+    type F = <C as GenericConfig<D>>::F;
+
+    let config = CircuitConfig::standard_recursion_config();
+
+    let mut builder = CircuitBuilder::<F, D>::new(config);
+
+    let x = builder.add_virtual_target();
+    let x_squared = builder.square(x);
+
+    builder.register_public_input(x);
```

Comment:
I think this example may be more meaningful if we keep `x` private, so we're proving `I know a square root of n` rather than `m is a square root of n` (which seems less interesting since sending `m, n` would prove the same statement)

Reply:
Ah okay!